contracts/instance/InstanceAccessManager.sol

```diff
@@ -22,201 +25,342 @@ contract InstanceAccessManager is
+        if (!inRole) {
+            revert IAccess.ErrorIAccessCallerIsNotRoleAdmin(_msgSender(), roleId);
+        }
+        _;
+    }
 
-    function initialize(address initialAdmin) external initializer
+    function initialize(address initialAdmin, address registry) external initializer 
     {
+        require(initialAdmin != address(0));
+        require(registry != address(0));
+
         // if size of the contract gets too large, this can be externalized which will reduce the contract size considerably
         _accessManager = new AccessManager(address(this));
-        // this service required admin rights to access manager to be able to grant/revoke roles
-        _accessManager.grantRole(_accessManager.ADMIN_ROLE(), initialAdmin, 0);
 
         __AccessManaged_init(address(_accessManager));
 
-        _createRole(RoleIdLib.toRoleId(_accessManager.ADMIN_ROLE()), ADMIN_ROLE_NAME, false, false);
-        _createRole(RoleIdLib.toRoleId(_accessManager.PUBLIC_ROLE()), PUBLIC_ROLE_NAME, false, false);
+        _registry = IRegistry(registry);
+        _idNext = CUSTOM_ROLE_ID_MIN;
+
+        _createRole(ADMIN_ROLE(), ADMIN_ROLE_NAME, IAccess.Type.Core);
+        _createRole(PUBLIC_ROLE(), PUBLIC_ROLE_NAME, IAccess.Type.Core);
+
+        // assume initialAdmin is instance service which requires admin rights to access manager during instance cloning
+        _accessManager.grantRole(ADMIN_ROLE().toInt(), initialAdmin, 0);
+
+        EnumerableSet.add(_roleMembers[ADMIN_ROLE()], address(this));
+        EnumerableSet.add(_roleMembers[ADMIN_ROLE()], initialAdmin);
     }
 
     //--- Role ------------------------------------------------------//
-    function createGifRole(RoleId roleId, string memory name) external restricted() {
-        _createRole(roleId, name, false, true);
+    // INSTANCE_SERVICE_ROLE 
+    // assume core roles are never revoked or renounced -> core roles admin is never active after intialization
+    function createCoreRole(RoleId roleId, string memory name, RoleId admin) 
+        external
+        restricted()
+    {
+        _validateRoleParameters(roleId, name, IAccess.Type.Core);// TODO put inside _createRole
+        _createRole(roleId, name, IAccess.Type.Core);
+        setRoleAdmin(roleId, admin);
+    }
+    // INSTANCE_SERVICE_ROLE
+    // assume gif roles can be revoked or renounced
+    function createGifRole(RoleId roleId, string memory name, RoleId admin) 
+        external
+        restricted()
+    {
+        _validateRoleParameters(roleId, name, IAccess.Type.Gif);
+        _createRole(roleId, name, IAccess.Type.Gif);
+        setRoleAdmin(roleId, admin);
     }
 
-    function createRole(RoleId roleId, string memory name) external restricted() {
-        _createRole(roleId, name, true, true);
+    // INSTANCE_OWNER_ROLE
+    // creates custom roles only
+    // TODO INSTANCE_OWNER_ROLE as default admin
+    function createCustomRole(string memory name, RoleId admin) 
```

Comment:
as core/gif roles are internal, it would probably be simpler to just call this `createRole`
```suggestion
    function createRole(string memory name, RoleId admin) 
```

Reply:
Done